esda/geary_local_mv.py

```diff
@@ -154,13 +178,21 @@ def _crand(self, zvariables):
         nvars = self.variables.shape[0]
         Gs = np.zeros((self.n, self.permutations))
         prange = list(range(self.permutations))
-        k = self.w.max_neighbors + 1
+        k = (
+            self.w.max_neighbors + 1
+            if isinstance(self.w, weights.W)
+            else self.w.cardinalities.max() + 1
+        )
         nn = self.n - 1
         rids = np.array([np.random.permutation(nn)[0:k] for i in prange])
         ids = np.arange(self.w.n)
-        ido = self.w.id_order
-        w = [self.w.weights[ido[i]] for i in ids]
-        wc = [self.w.cardinalities[ido[i]] for i in ids]
+        if isinstance(self.w, weights.W):
```

Comment:
This is a nice illustration of gains from using the graph over W, in terms of simplification.